### ISSUE TITLE
bug in SBX implementation 

### ISSUE THREAD

**Issue by MLopez-Ibanez** [esa/pagmo2#421]
https://github.com/esa/pagmo2/blob/ab1f3a945a4b78886bb270532d80215f08782a3a/src/algorithms/nsga2.cpp#L439-L442

site1 and site2 above will be in `[0, Di-1]`, however, it should be in `[Dc, D]` as the chromosome positions are `[0, ..., Dc, Dc+1, ...., Dc+Di]`. 

Also, there are two nested loops and each inner may overwrite the whole children. That doesn't seem correct. I think it should be:

```c++
    // This implements two-point binary crossover and applies it to the integer part of the chromosome
    if (Di == 0)
        return;

    if (drng(m_e) < m_cr) {
        std::uniform_int_distribution<decltype(Dc)> ra_num(Dc, D - 1);
        auto site1 = ra_num(m_e);
        auto site2 = ra_num(m_e);
        if (site1 > site2) {
            std::swap(site1, site2);
        }
        auto j = Dc;
        for (; j < site1; ++j) {
            child1[j] = parent1[j];
            child2[j] = parent2[j];
        }
        for (; j < site2; ++j) {
            child1[j] = parent2[j];
            child2[j] = parent1[j];
        }
        for (; j < D; ++j) {
            child1[j] = parent1[j];
            child2[j] = parent2[j];
        }
    } else {
        for (auto i = Dc; i < D; ++i) {
            child1[i] = parent1[i];
            child2[i] = parent2[i];
        }
    }
```

**Comment by MLopez-Ibanez** [comment]
Another minor bug is that:
```c++
if (drng(m_e) <= m_cr)
```
is called two times, hence, if we have a mixed-integer problem, this means that the crossover rate per solution is doubled (there are two chances per crossover call that the solution will be modified). 

I think the code should do once and at the very start:
```c++
if (drng(m_e) >= m_cr) return
```

**Comment by MLopez-Ibanez** [comment]
Actually, since child1 is a copy of parent1 and child2 is a copy of parent2, then we can simply do:
```c++
    // This implements two-point binary crossover and applies it to the integer part of the chromosome
    if (Di == 0)
        return;

    if (drng(m_e) < m_cr) {
        std::uniform_int_distribution<decltype(Dc)> ra_num(Dc, D - 1);
        auto site1 = ra_num(m_e);
        auto site2 = ra_num(m_e);
        if (site1 > site2) {
            std::swap(site1, site2);
        }
        // child1 is a copy of parent1 and child2 is a copy of parent2, so just change the middle section.
        for (auto j = site1; j < site2; ++j) {
            child1[j] = parent2[j];
            child2[j] = parent1[j];
        }        
    } else {
        for (auto i = Dc; i < D; ++i) {
            child1[i] = parent1[i];
            child2[i] = parent2[i];
        }
    }
```

**Comment by darioizzo** [comment]
Yes the code makes no sense. Fixed in #422. Have a look if you spot anything off now.

**Comment by bluescarni** [comment]
Fixed by #422.
